### PR TITLE
Re-submission of Testnet and Scaling-Testnet (STN) integration

### PR DESCRIPTION
**Pull request by AustEcon**
@teran-mckinney and @joshua-s. This is a re-submission of #24 

Thank you very much for your thoughtful feedback. I have made some revisions.

The design decisions I am representing here is that of:
1) Keeping the `NetworkAPI `class *only* for the "bread and butter" functions (getting balance/utxos/transactions/sending) - called from `PrivateKey `/ `PrivateKeyTest`.
2) Normalizing the various APIs in `NetworkAPI `for this purpose (e.g. override names of functions or return format etc. so that they all end up being uniform for these "bread and butter functions".
3) Placing integrated `BIP32 `support on the back-burner for now (tends to require api_key). Would require discussion around how best to do it and is probably best done in an entirely separate class or repo (leaving `PrivateKey `and `PrivateKeyTest `+ `NetworkAPI `well alone)

Other changes (as per #24 ):
- Use `NetworkAPI `as an instance object --> easy combining of the main / test / stn network function calls quite easily (but if it is better kept as a `@classmethod`s please do let me know)
- Changed behavior of the `send_op_return `and `create_op_return `functions to **not consolidate utxos** by default (very annoying for bcat:// where you need to have numerous confirmed utxos on standby). Added them to `PrivateKeyTest `. Also added `outputs `and `leftover `fields for more granular control from these functions.
- Added tests to `test_services.py` for main, test and stn networks.
- Other minor fixes mentioned in #24 

PS: I'm not sure if changing NetworkAPI over to creating instance objects is a good way to go or not. I am eager to get feedback re: best way to implement this all... pros and cons. I think the least controversial way would probably be to go similar to the main "bit" (BTC) library and laboriously write out main, test and stn separately and use as classmethods... and then add to that my idea of redundancy handling with deque.rotate()... but before going that route I wanted to explore if we can do better...

I would like to cut a new release soon. Once this is sorted + docs tidied up. Things will be looking pretty nice :) The next main improvement would be the addition of another api... Blockchair is giving me grief though... I think I'll park that and wait for Whatsonchain...

### CODE REVIEW

File: bitsv/network/services/network.py [hunk 1]
```diff
@@ -10,6 +11,34 @@ def set_service_timeout(seconds):
...
+            'Accept': 'application/json',
+        }
+
+    def _get_authorized_headers(self):
+        headers = self._get_headers()
+        headers['api_key'] = self.api_key
+        return headers
+
+    def get_transactions(self, address):
+        # override "get_transactions" namespace to return a list of txids only to keep all apis the same
```
Comment (AustEcon) [inline bitsv/network/services/network.py]: The `get_balance `function returns a list of txids and is far more performant than the `get_transactions` function - especially for keys that have a voluminous hx of spamming op_return data... I want the NetworkAPI bread and butter functions to be quick and responsive... once you have the txids you can always pick and choose via the `get_transaction` function.

**Comment by teran-mckinney** [comment]
Hey,

I've looked over this and at first glance I think it looks good. I want to test and see if the changes break bitcoinacceptor-python for me. So I'm trying to install, but the bitcoinx requirement requires Python 3.6. I am on Debian 9 which uses Python 3.5 by default.

Were you planning to break Python 3.5 support? What is bitcoinx for?

Thanks,

Teran

**Comment by AustEcon** [comment]
Okay cool. So you're okay with changing NetworkAPI over to instance methods instead? I'm actually discussing with Josh as we speak.
- Hmmmm... BitcoinX is a library being built out by kyuupichan, the dev behind ElectrumX and ElectrumSV and I added it as a dependency for signing messages and BIP32 support (experimental)... I can use it to sign messages in a way that spits out a signature compatible with ElectrumSV and I can do path derivations for handcash, electrumsv and basically any other kind of commercial wallet I want now through the bip32.py methods... but breaking python3.5 is a bummer... I'll have to think about this one... maybe bip32 stuff should be in a new repo from bitsv?

**Comment by teran-mckinney** [comment]
> Okay cool. So you're okay with changing NetworkAPI over to instance methods instead? I'm actually discussing with Josh as we speak.
> 
>     * Hmmmm... BitcoinX is a library being built out by kyuupichan, the dev behind ElectrumX and ElectrumSV and I added it as a dependency for signing messages and BIP32 support (experimental)... I can use it to sign messages in a way that spits out a signature compatible with ElectrumSV and I can do path derivations for handcash, electrumsv and basically any other kind of commercial wallet I want now through the bip32.py methods... but breaking python3.5 is a bummer... I'll have to think about this one... maybe bip32 stuff should be in a new repo from bitsv?

bip32 is pretty cool but I tend to think it's better separate if possible. If we need to make bitsv easier to use to have  bip32 addon, I think that is fine. Or maybe it works well as it is that way. I think bip32 is more of a wallet feature and someone building on bitsv would have to roll on a lot of those features.

If the bip32 library could tap into bit, bitcash, and bitsv, that would be really awesome.

**Comment by teran-mckinney** [comment]
Ok, another question. For the bitindex sorting, is that by unspent size or by confirmations? I am used to APIs returning sorted by confirmations but maybe I shouldn't be relying on that.

**Comment by AustEcon** [comment]
Good question. I don't know - It looks like if you pass "None" to sort it returns in ascending order of amount and ascending order of confirmations...

This raises a question... should we be enforcing in the NetworkAPI.get_unspents() the sorting of utxos?? It could reduce performance for large utxos sets to do big sorting operations though...

May be best to make no guarantees about the ordering and leave that to the programmer / local machine to sort as needed?

I actually have some basic utxo sorting functions that I was going to add bitsv.utils (as a little thing) by the way...

File: bitsv/wallet.py [hunk 2]
```diff
@@ -148,7 +154,7 @@ def __init__(self, wif=None):
         self.unspents = []
         self.transactions = []
         self.network = 'main'
```
Comment (AustEcon) [inline bitsv/wallet.py]: You may wonder why I have still kept the self.network = 'main' (not used)... it is because it may become used in next PR...

Comment (AustEcon) [inline bitsv/wallet.py]: You may wonder why I have still kept the self.network = 'main' (not used)... it is because it may become used in next PR...

**Comment by AustEcon** [comment]
For *this PR*, I would rather not mess about with the bip32 stuff to reclaim python3.5 support. But it is definately something I am happy to look at for another PR e.g. moving bip32 stuff to another repo that extends bitsv (and can then add bitcoinx as a dependency to that)... Keeping python3.5 and pypy functioning *is* important to me. Especially seeing as though ubuntu defaults to python3.5 and in my experience it is a pain in the arse to properly configure a new python without breaking stuff... haha

**Comment by teran-mckinney** [comment]
If you can, let me know before you push a new release up to pypi so I can test it with walkingliberty and bitcoinacceptor to see if I need to pin anything back or make adjustments.

**Comment by AustEcon** [comment]
Okay awesome! Will do. Thanks!

**Pull request merged**